### ISSUE TITLE
[EPCO-50] Model will predict sets of possible states

### ISSUE THREAD

**Issue by hatfield-c** [hatfield-c/epco#50]
The previous attempts have failed to provide good behavior for control of the mujoco robot. I believe these failures are due to the fact that previous attempts would work well in small, localized area, but large amounts of the problem space would be uncontained by the voxels. I attempted various methods to interpolate between the voxels, but this has proven to be insufficient.

I have returned to the drawing board, and considered what it is we're trying to do here. Principally, we wanted to learn a formulation of knowledge that was compatible with planner. The push towards XAI was in pursuit of that goal, as it was reasoned that if the AI is explainable then it can be utilized by a planner. However, as mentioned the previous attempts at this fail at properly capturing the data distribution when dimensionality becomes too high.

As such, I've developed a new mathematical object similar to the voxel. I call it the vubo, and it is functionally a series of spheres that are stretched from the origin. Essentially we will train a traditional neural network to predict the radii of these spheres, such that when the robot dog is in state s then the spheres tell us what changes to s are possible from this state. Thusly, we can know exactly the full extent of reachable states (as approximated by the vubo) and feed this information into a planner. A second neural network will then be generated which learns what actions at a state cause what results. A planner can then generate a plan of states it wishes to reach, and the executer will determine what actions to take at each step to achieve the desired results.

**Comment by hatfield-c** [comment]
There unfortunately are some issues with the training of the vubos.

Currently they are being trained to perform binary discrimination between points that are contained within the vubo, and thus reachable, and points that are not contained in the vubo and are thus unreachable.

This issue seems to arise from my use of binary cross entropy and a sigmoid classification layer as the means of training. Because we are working in a normalized environment where everything is between -1 and 1, the sigmoid layer has trouble distinguishing between positive and negative points. Especially since the input space is the distance of a query from a vubo center, the values tend to not be very far from 0, which causes most points to have sigmoidal values very close to 0.5, even if they are ostensibly supposed to be calculated as far away. We could try scaling the input into the sigmoid function, but this causes instability with the gradient updates. Neural networks don't usually have this problem as they can scale their output data to very large values.

**Comment by hatfield-c** [comment]
Another issue with vubos has arisen - the lack of negative data points to compare against.

We can only learn from the positive data points gathered by taking random actions at a particular state, but it is impossible for us to generate counter-examples.

As such, I am exploring the use of one-class classification systems. While primitive, k-means might actually work well here.
